query/src/main/java/org/infinispan/query/impl/massindex/AbstractIndexWorker.java

```diff
@@ -0,0 +1,95 @@
+import org.infinispan.commons.util.Util;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.distexec.DistributedCallable;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.filter.AcceptAllKeyValueFilter;
+import org.infinispan.filter.KeyValueFilter;
+import org.infinispan.interceptors.locking.ClusteringDependentLogic;
+import org.infinispan.iteration.impl.EntryRetriever;
+import org.infinispan.marshall.core.MarshalledValue;
+import org.infinispan.metadata.Metadata;
+
+import java.util.Set;
+
+/**
+ * Base class for mass indexer tasks.
+ *
+ * @author gustavonalle
+ * @since 7.1
+ */
+public abstract class AbstractIndexWorker implements DistributedCallable<Void, Void, Void> {
+
+   protected Cache<?, ?> cache;
+   protected final Class<?> entity;
+   protected final String indexName;
+   protected IndexUpdater indexUpdater;
+
+   private ClusteringDependentLogic clusteringDependentLogic;
+   private EntryRetriever entryRetriever;
+
+   public AbstractIndexWorker(Class<?> entity, String indexName) {
+      this.entity = entity;
+      this.indexName = indexName;
+   }
+
+   @Override
+   public void setEnvironment(Cache<Void, Void> cache, Set<Void> inputKeys) {
+      this.cache = cache;
+      this.indexUpdater = new IndexUpdater(cache);
+      ComponentRegistry componentRegistry = cache.getAdvancedCache().getComponentRegistry();
+      this.entryRetriever = componentRegistry.getComponent(EntryRetriever.class);
+      this.clusteringDependentLogic = componentRegistry.getComponent(ClusteringDependentLogic.class);
+   }
+
+   protected abstract void preIndex();
+
+   protected abstract void postIndex();
+
+   private KeyValueFilter getFilter() {
+      boolean replicated = cache.getCacheConfiguration().clustering().cacheMode().isReplicated();
+      return replicated ? AcceptAllKeyValueFilter.getInstance() : new PrimaryOwnersKeyValueFilter(clusteringDependentLogic);
+   }
+
+   private Object extractValue(Object wrappedValue) {
+      if (wrappedValue instanceof MarshalledValue)
+         return ((MarshalledValue) wrappedValue).get();
+      return wrappedValue;
+   }
+
+   @Override
+   @SuppressWarnings("unchecked")
+   public Void call() throws Exception {
+      preIndex();
+      KeyValueFilter filter = getFilter();
+      try (CloseableIterator<CacheEntry<Object, String>> iterator = entryRetriever.retrieveEntries(filter, null, Util.asSet(Flag.CACHE_MODE_LOCAL), null)) {
+         while (iterator.hasNext()) {
+            CacheEntry<Object, String> next = iterator.next();
+            Object value = extractValue(next.getValue());
+            if (value != null && value.getClass().equals(entity))
+               indexUpdater.updateIndex(next.getKey(), value, indexName);
+         }
+      }
+      postIndex();
+      return null;
+   }
+
+
+   private class PrimaryOwnersKeyValueFilter implements KeyValueFilter {
```

Comment:
This class could be static or it could simply be non-static but then it does not need the ```cdl``` field as that one is available in the outer class.